### PR TITLE
occamy: Add basic FPGA flow

### PR DESCRIPTION
**Pull request by niwis**
This adds a basic Vivado flow for a VCU128 FPGA emulation platform. Occamy is packaged as a Xilinx IP in `occamy_xilinx`. `occamy_xilinx` is instantiated in the block design of `occamy_vcu128` together with the following components:
* A Xilinx AXI SmartConnect, including an AXI crossbar, protocol converters, data width converters, clock converters,
* An 8GB HBM with 8 AXI slave ports (`0x10_0000_0000` - `0x12_0000_0000`, `hbi_x`), each addressing 4 memory channels of the HBM (2 additional AXI slaves realise aliasing of `0x8000_0000` - `0x1_0000_0000`, `hbi_0` and `hbi_1`, to the first 8 memory channels), and 2 APB ports for configuration (`0x7040_0000` - `0x7080_0000` and `0x7080_0000` - `0x70C0_0000`, `out_pcie`),
* A Xilinx AXI UART 16550 (`0x7000_0000` - `0x7040_0000`, `out_pcie`)

Running `make fpga` from `hw/system/occamy` generates a bitstream.

### CODE REVIEW

File: hw/system/occamy/fpga/occamy_vcu128_bd.tcl [hunk 1]
```diff
@@ -0,0 +1,443 @@
+# Copyright 2020 ETH Zurich and University of Bologna.
+# Solderpad Hardware License, Version 0.51, see LICENSE for details.
+# SPDX-License-Identifier: SHL-0.51
+#
+# Nils Wistoff <nwistoff@iis.ee.ethz.ch>
+#
+# Block design generation script generated by Vivado.
```
Comment (niwis) [inline hw/system/occamy/fpga/occamy_vcu128_bd.tcl:7]: Is the license header ok for a script generated by a commercial tool?

File: hw/system/occamy/fpga/occamy_vcu128.tcl [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+generate_target all [get_files ./occamy_vcu128/occamy_vcu128.srcs/sources_1/bd/occamy_vcu128/occamy_vcu128.bd]
+export_ip_user_files -of_objects [get_files ./occamy_vcu128/occamy_vcu128.srcs/sources_1/bd/occamy_vcu128/occamy_vcu128.bd] -no_script -sync -force -quiet
+create_ip_run [get_files -of_objects [get_fileset sources_1] ./occamy_vcu128/occamy_vcu128.srcs/sources_1/bd/occamy_vcu128/occamy_vcu128.bd]
+
+# Re-add occamy includes
+export_ip_user_files -of_objects [get_ips occamy_vcu128_occamy_xilinx_0_0] -no_script -sync -force -quiet
+eval [exec sed {s/current_fileset/get_filesets occamy_vcu128_occamy_xilinx_0_0/} define_defines_includes_no_simset.tcl]
+
+# Do NOT insert BUFGs on high-fanout nets (e.g. reset). This will backfire during placement.
+set_param logicopt.enableBUFGinsertHFN no
```
Comment (zarubaf) [inline hw/system/occamy/fpga/occamy_vcu128.tcl:40]: Crazy tool...

File: hw/system/occamy/fpga/occamy_vcu128_bd.tcl (hunk 1 above)
Comment (zarubaf) [reply hw/system/occamy/fpga/occamy_vcu128_bd.tcl:7]: Absolutely, that is perfect :-)

Btw, how do you generate the BD? Should we maybe add this to a README or something in the `docs` folder? Like how would I change the BD in the GUI and re-generate the `tcl` script? I guess the license header should also be added manually? If that is a problem we can generate an exception in the license checker for this automatically generated file :-)

File: hw/system/occamy/fpga/vivado_ips/occamy_xilinx.tcl [hunk 3]
```diff
@@ -0,0 +1,31 @@
+# Copyright 2020 ETH Zurich and University of Bologna.
```
Comment (zarubaf) [inline hw/system/occamy/fpga/vivado_ips/occamy_xilinx.tcl:1]: I really like this!

**Review by zarubaf — APPROVED** [review]
@niwis that is really fantastic. Super clean, no hacky top-level `ifdefs` <3 

I hope that we get this running without doing too much (any) top-level simulation ;-)

File: hw/system/occamy/fpga/occamy_vcu128_bd.tcl (hunk 1 above)
Comment (niwis) [reply hw/system/occamy/fpga/occamy_vcu128_bd.tcl:7]: With `occamy_vcu128.xpr` open in the Vivado GUI, the block design can be accessed on the left (`Open Block Design`). There, the block design can be viewed and modified. The `tcl` script for the block design can then by generated via `File`->`Export`->`Export Block Design`. There are some manual adjustments to this file, perhaps I should try to move these to the main script (`occamy_vcu128.tcl`). Usually I `diff` the exported with the original version of `occamy_vcu128_bd.tcl`.

I'll gladly put this info into the `docs` folder  - not very sure where, though.

**Pull request merged**